File: go/hybrid/hybrid_kem.go
// Package hybrid implements Hybrid KEM schemes combining classical
// ECDH key exchange with ML-KEM for post-quantum/classical security.
//
// The combiner ensures security holds if either the classical or
// post-quantum component remains secure.
//
// Supported schemes:
//   - X25519 + ML-KEM-768 (IETF standard hybrid for TLS)
//   - ECDH-P256 + ML-KEM-768
//   - X25519 + ML-KEM-1024 (higher security)
//   - ECDH-P384 + ML-KEM-1024
//
// KDF: SHA3-256(ss_classical || ss_pq || label)
package hybrid

import (
	"crypto/ecdh"
	"crypto/rand"
	"errors"
	"io"

	"golang.org/x/crypto/sha3"

	"github.com/baron-chain/PQC-Standards-Implementation/go/internal/params"
	"github.com/baron-chain/PQC-Standards-Implementation/go/mlkem"
)

// Scheme identifies a hybrid KEM scheme.
type Scheme struct {
	Name           string
	ClassicalCurve ecdh.Curve
	MLKEMParams    params.ParameterSet
	Label          []byte
}

var (
	// X25519MlKem768 combines X25519 with ML-KEM-768.
	X25519MlKem768 = Scheme{
		Name:           "X25519-MLKEM768",
		ClassicalCurve: ecdh.X25519(),
		MLKEMParams:    params.MlKem768,
		Label:          []byte("X25519-MLKEM768"),
	}

	// EcdhP256MlKem768 combines ECDH-P256 with ML-KEM-768.
	EcdhP256MlKem768 = Scheme{
		Name:           "ECDHP256-MLKEM768",
		ClassicalCurve: ecdh.P256(),
		MLKEMParams:    params.MlKem768,
		Label:          []byte("ECDHP256-MLKEM768"),
	}

	// X25519MlKem1024 combines X25519 with ML-KEM-1024.
	X25519MlKem1024 = Scheme{
		Name:           "X25519-MLKEM1024",
		ClassicalCurve: ecdh.X25519(),
		MLKEMParams:    params.MlKem1024,
		Label:          []byte("X25519-MLKEM1024"),
	}

	// EcdhP384MlKem1024 combines ECDH-P384 with ML-KEM-1024.
	EcdhP384MlKem1024 = Scheme{
		Name:           "ECDHP384-MLKEM1024",
		ClassicalCurve: ecdh.P384(),
		MLKEMParams:    params.MlKem1024,
		Label:          []byte("ECDHP384-MLKEM1024"),
	}
)

// HybridKeyPair holds a hybrid key pair.
type HybridKeyPair struct {
	// EK is the combined encapsulation key (classical_pk || pq_ek).
	EK []byte
	// DK is the combined decapsulation key (classical_sk || pq_dk).
	DK []byte
	// ClassicalEKSize is the size of the classical public key portion.
	ClassicalEKSize int
	// ClassicalDKSize is the size of the classical secret key portion.
	ClassicalDKSize int
}

// EncapsResult holds the result of hybrid encapsulation.
type EncapsResult struct {
	// SharedSecret is the combined 32-byte shared secret.
	SharedSecret [32]byte
	// Ciphertext is the combined ciphertext (classical_ct || pq_ct).
	Ciphertext []byte
	// ClassicalCTSize is the size of the classical ciphertext portion.
	ClassicalCTSize int
}

// combineSecrets derives the hybrid shared secret using SHA3-256.
func combineSecrets(ssClassical, ssPQ, label []byte) [32]byte {
	h := sha3.New256()
	h.Write(ssClassical)
	h.Write(ssPQ)
	h.Write(label)
	var out [32]byte
	copy(out[:], h.Sum(nil))
	return out
}

// KeyGen generates a hybrid key pair using the specified scheme.
func KeyGen(scheme Scheme, rng io.Reader) (*HybridKeyPair, error) {
	if rng == nil {
		rng = rand.Reader
	}

	// Classical key generation
	classicalSK, err := scheme.ClassicalCurve.GenerateKey(rng)
	if err != nil {
		return nil, errors.New("hybrid: classical keygen failed: " + err.Error())
	}
	classicalPK := classicalSK.PublicKey().Bytes()
	classicalSKBytes := classicalSK.Bytes()

	// ML-KEM key generation
	pqEK, pqDK, err := mlkem.KeyGen(scheme.MLKEMParams, rng)
	if err != nil {
		return nil, errors.New("hybrid: ML-KEM keygen failed: " + err.Error())
	}

	// Combine keys
	ek := make([]byte, 0, len(classicalPK)+len(pqEK))
	ek = append(ek, classicalPK...)
	ek = append(ek, pqEK...)

	dk := make([]byte, 0, len(classicalSKBytes)+len(pqDK))
	dk = append(dk, classicalSKBytes...)
	dk = append(dk, pqDK...)

	return &HybridKeyPair{
		EK:              ek,
		DK:              dk,
		ClassicalEKSize: len(classicalPK),
		ClassicalDKSize: len(classicalSKBytes),
	}, nil
}

// Encaps encapsulates using the hybrid scheme, producing a shared secret and ciphertext.
func Encaps(scheme Scheme, ek []byte, classicalEKSize int, rng io.Reader) (*EncapsResult, error) {
	if rng == nil {
		rng = rand.Reader
	}

	classicalPKBytes := ek[:classicalEKSize]
	pqEK := ek[classicalEKSize:]

	// Parse the classical public key
	classicalPK, err := scheme.ClassicalCurve.NewPublicKey(classicalPKBytes)
	if err != nil {
		return nil, errors.New("hybrid: invalid classical public key: " + err.Error())
	}

	// Classical ECDH encapsulation: generate ephemeral key pair, DH
	ephSK, err := scheme.ClassicalCurve.GenerateKey(rng)
	if err != nil {
		return nil, errors.New("hybrid: classical ephemeral keygen failed: " + err.Error())
	}
	ssClassical, err := ephSK.ECDH(classicalPK)
	if err != nil {
		return nil, errors.New("hybrid: classical ECDH failed: " + err.Error())
	}
	ctClassical := ephSK.PublicKey().Bytes()

	// ML-KEM encapsulation
	ssPQ, ctPQ, err := mlkem.Encapsulate(scheme.MLKEMParams, pqEK, rng)
	if err != nil {
		return nil, errors.New("hybrid: ML-KEM encaps failed: " + err.Error())
	}

	combinedSS := combineSecrets(ssClassical, ssPQ[:], scheme.Label)

	ct := make([]byte, 0, len(ctClassical)+len(ctPQ))
	ct = append(ct, ctClassical...)
	ct = append(ct, ctPQ...)

	return &EncapsResult{
		SharedSecret:    combinedSS,
		Ciphertext:      ct,
		ClassicalCTSize: len(ctClassical),
	}, nil
}

// Decaps decapsulates using the hybrid scheme, recovering the shared secret.
func Decaps(scheme Scheme, dk, ct []byte, classicalDKSize, classicalCTSize int) ([32]byte, error) {
	classicalSKBytes := dk[:classicalDKSize]
	pqDK := dk[classicalDKSize:]

	ctClassical := ct[:classicalCTSize]
	ctPQ := ct[classicalCTSize:]

	// Parse the classical secret key
	classicalSK, err := scheme.ClassicalCurve.NewPrivateKey(classicalSKBytes)
	if err != nil {
		return [32]byte{}, errors.New("hybrid: invalid classical secret key: " + err.Error())
	}

	// Parse the ephemeral public key from ciphertext
	ephPK, err := scheme.ClassicalCurve.NewPublicKey(ctClassical)
	if err != nil {
		return [32]byte{}, errors.New("hybrid: invalid classical ciphertext: " + err.Error())
	}

	// Classical ECDH decapsulation
	ssClassical, err := classicalSK.ECDH(ephPK)
	if err != nil {
		return [32]byte{}, errors.New("hybrid: classical ECDH decaps failed: " + err.Error())
	}

	// ML-KEM decapsulation
	ssPQ := mlkem.Decapsulate(scheme.MLKEMParams, pqDK, ctPQ)

	combinedSS := combineSecrets(ssClassical, ssPQ[:], scheme.Label)
	return combinedSS, nil
}

File: python/hybrid/hybrid_kem.py
"""Hybrid KEM: combining classical ECDH with ML-KEM.

Hybrid KEMs ensure security holds if either the classical or
post-quantum component remains secure.

Supported schemes:
- X25519 + ML-KEM-768 (IETF standard hybrid for TLS)
- ECDH-P256 + ML-KEM-768
- X25519 + ML-KEM-1024
- ECDH-P384 + ML-KEM-1024

KDF: SHA3-256(ss_classical || ss_pq || label)
"""

import hashlib
from dataclasses import dataclass
from typing import Optional

import sys
import os

# Add parent directory to path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from mlkem.kem import keygen as mlkem_keygen, encaps as mlkem_encaps, decaps as mlkem_decaps
from mlkem.params import ML_KEM_768, ML_KEM_1024, MLKEMParams
from hybrid.x25519 import x25519_keygen, x25519_encaps, x25519_decaps


def _sha3_256(*inputs: bytes) -> bytes:
    """SHA3-256 of concatenated inputs."""
    h = hashlib.sha3_256()
    for inp in inputs:
        h.update(inp)
    return h.digest()


def _combine_secrets(ss_classical: bytes, ss_pq: bytes, label: bytes) -> bytes:
    """Combine shared secrets: SHA3-256(ss_classical || ss_pq || label)."""
    return _sha3_256(ss_classical, ss_pq, label)


@dataclass(frozen=True)
class HybridScheme:
    """Definition of a hybrid KEM scheme."""
    name: str
    label: bytes
    classical_type: str  # "x25519", "P-256", or "P-384"
    mlkem_params: MLKEMParams


@dataclass
class HybridKeyPair:
    """Hybrid key pair."""
    ek: bytes
    dk: bytes
    classical_ek_size: int
    classical_dk_size: int


@dataclass
class HybridEncapsResult:
    """Result of hybrid encapsulation."""
    shared_secret: bytes
    ciphertext: bytes
    classical_ct_size: int


# Scheme definitions
X25519_MLKEM768 = HybridScheme(
    name="X25519-MLKEM768",
    label=b"X25519-MLKEM768",
    classical_type="x25519",
    mlkem_params=ML_KEM_768,
)

ECDHP256_MLKEM768 = HybridScheme(
    name="ECDHP256-MLKEM768",
    label=b"ECDHP256-MLKEM768",
    classical_type="P-256",
    mlkem_params=ML_KEM_768,
)

X25519_MLKEM1024 = HybridScheme(
    name="X25519-MLKEM1024",
    label=b"X25519-MLKEM1024",
    classical_type="x25519",
    mlkem_params=ML_KEM_1024,
)

ECDHP384_MLKEM1024 = HybridScheme(
    name="ECDHP384-MLKEM1024",
    label=b"ECDHP384-MLKEM1024",
    classical_type="P-384",
    mlkem_params=ML_KEM_1024,
)


def _classical_keygen(classical_type: str) -> tuple[bytes, bytes]:
    """Generate a classical key pair."""
    if classical_type == "x25519":
        return x25519_keygen()
    else:
        from hybrid.ecdh_nist import ecdh_keygen
        pk, sk, _ = ecdh_keygen(classical_type)
        return pk, sk


def _classical_encaps(classical_type: str, peer_pk: bytes) -> tuple[bytes, bytes]:
    """Classical KEM-style encapsulation."""
    if classical_type == "x25519":
        return x25519_encaps(peer_pk)
    else:
        from hybrid.ecdh_nist import ecdh_encaps
        return ecdh_encaps(classical_type, peer_pk)


def _classical_decaps(classical_type: str, sk: bytes, ct: bytes) -> bytes:
    """Classical KEM-style decapsulation."""
    if classical_type == "x25519":
        return x25519_decaps(sk, ct)
    else:
        from hybrid.ecdh_nist import ecdh_decaps
        return ecdh_decaps(classical_type, sk, ct)


def hybrid_keygen(scheme: HybridScheme) -> HybridKeyPair:
    """Generate a hybrid key pair.

    Args:
        scheme: The hybrid KEM scheme to use.

    Returns:
        HybridKeyPair with combined keys.
    """
    classical_pk, classical_sk = _classical_keygen(scheme.classical_type)
    pq_ek, pq_dk = mlkem_keygen(scheme.mlkem_params)

    ek = classical_pk + pq_ek
    dk = classical_sk + pq_dk

    return HybridKeyPair(
        ek=ek,
        dk=dk,
        classical_ek_size=len(classical_pk),
        classical_dk_size=len(classical_sk),
    )


def hybrid_encaps(scheme: HybridScheme, ek: bytes, classical_ek_size: int) -> HybridEncapsResult:
    """Hybrid encapsulation.

    Args:
        scheme: The hybrid KEM scheme.
        ek: Combined encapsulation key.
        classical_ek_size: Size of the classical portion.

    Returns:
        HybridEncapsResult with shared secret and ciphertext.
    """
    classical_pk = ek[:classical_ek_size]
    pq_ek = ek[classical_ek_size:]

    ss_classical, ct_classical = _classical_encaps(scheme.classical_type, classical_pk)
    ss_pq, ct_pq = mlkem_encaps(pq_ek, scheme.mlkem_params)

    combined_ss = _combine_secrets(ss_classical, ss_pq, scheme.label)

    return HybridEncapsResult(
        shared_secret=combined_ss,
        ciphertext=ct_classical + ct_pq,
        classical_ct_size=len(ct_classical),
    )


def hybrid_decaps(
    scheme: HybridScheme,
    dk: bytes,
    ct: bytes,
    classical_dk_size: int,
    classical_ct_size: int,
) -> bytes:
    """Hybrid decapsulation.

    Args:
        scheme: The hybrid KEM scheme.
        dk: Combined decapsulation key.
        ct: Combined ciphertext.
        classical_dk_size: Size of the classical secret key portion.
        classical_ct_size: Size of the classical ciphertext portion.

    Returns:
        32-byte combined shared secret.
    """
    classical_sk = dk[:classical_dk_size]
    pq_dk = dk[classical_dk_size:]

    ct_classical = ct[:classical_ct_size]
    ct_pq = ct[classical_ct_size:]

    ss_classical = _classical_decaps(scheme.classical_type, classical_sk, ct_classical)
    ss_pq = mlkem_decaps(pq_dk, ct_pq, scheme.mlkem_params)

    return _combine_secrets(ss_classical, ss_pq, scheme.label)
